pkg/compactor/blocks_cleaner_test.go

```diff
@@ -406,6 +410,231 @@ func TestBlocksCleaner_ShouldRemoveMetricsForTenantsNotBelongingAnymoreToTheShar
+		return time.Now().Add(time.Duration(hours)*time.Hour).Unix() * 1000
+	}
+
+	block1 := createTSDBBlock(t, bucketClient, "user-1", ts(-10), ts(-8), nil)
+	block2 := createTSDBBlock(t, bucketClient, "user-1", ts(-8), ts(-6), nil)
+	block3 := createTSDBBlock(t, bucketClient, "user-2", ts(-10), ts(-8), nil)
+	block4 := createTSDBBlock(t, bucketClient, "user-2", ts(-8), ts(-6), nil)
+
+	cfg := BlocksCleanerConfig{
+		DeletionDelay:      time.Hour,
+		CleanupInterval:    time.Minute,
+		CleanupConcurrency: 1,
+	}
+
+	ctx := context.Background()
+	logger := log.NewNopLogger()
+	reg := prometheus.NewPedanticRegistry()
+	scanner := tsdb.NewUsersScanner(bucketClient, tsdb.AllUsers, logger)
+	cfgProvider := newMockConfigProvider()
+
+	cleaner := NewBlocksCleaner(cfg, bucketClient, scanner, cfgProvider, logger, reg)
+
+	assertBlocksExist := func(exists1, exists2, exists3, exists4 bool) {
+		exists, err := bucketClient.Exists(ctx, path.Join("user-1", block1.String(), metadata.MetaFilename))
+		require.NoError(t, err)
+		assert.Equal(t, exists1, exists)
+
+		exists, err = bucketClient.Exists(ctx, path.Join("user-1", block2.String(), metadata.MetaFilename))
+		require.NoError(t, err)
+		assert.Equal(t, exists2, exists)
+
+		exists, err = bucketClient.Exists(ctx, path.Join("user-2", block3.String(), metadata.MetaFilename))
+		require.NoError(t, err)
+		assert.Equal(t, exists3, exists)
+
+		exists, err = bucketClient.Exists(ctx, path.Join("user-2", block4.String(), metadata.MetaFilename))
+		require.NoError(t, err)
+		assert.Equal(t, exists4, exists)
+	}
+
+	require.NoError(t, cleaner.cleanUsers(ctx, true))
+	assertBlocksExist(true, true, true, true)
+
+	// Existing behaviour - retention period disabled.
+
+	cfgProvider.UserRetentionPeriods["user-1"] = 0
+	cfgProvider.UserRetentionPeriods["user-2"] = 0
+
+	require.NoError(t, cleaner.cleanUsers(ctx, false))
+	assertBlocksExist(true, true, true, true)
+
+	assert.NoError(t, prom_testutil.GatherAndCompare(reg, strings.NewReader(`
+		# HELP cortex_bucket_blocks_count Total number of blocks in the bucket. Includes blocks marked for deletion, but not partial blocks.
+		# TYPE cortex_bucket_blocks_count gauge
+		cortex_bucket_blocks_count{user="user-1"} 2
+		cortex_bucket_blocks_count{user="user-2"} 2
+		# HELP cortex_bucket_blocks_marked_for_deletion_count Total number of blocks marked for deletion in the bucket.
+		# TYPE cortex_bucket_blocks_marked_for_deletion_count gauge
+		cortex_bucket_blocks_marked_for_deletion_count{user="user-1"} 0
+		cortex_bucket_blocks_marked_for_deletion_count{user="user-2"} 0
+		# HELP cortex_compactor_blocks_marked_for_deletion_total Total number of blocks marked for deletion in compactor.
+		# TYPE cortex_compactor_blocks_marked_for_deletion_total counter
+		cortex_compactor_blocks_marked_for_deletion_total{reason="retention"} 0
+	`),
+		"cortex_bucket_blocks_count",
+		"cortex_bucket_blocks_marked_for_deletion_count",
+		"cortex_compactor_blocks_marked_for_deletion_total",
+	))
```

Comment:
Showing as an example here. When you have multiple tests in the same function and you're not using the testing table approach, then may be easier to read wrapping into blocks like:

```suggestion
{
	cfgProvider.UserRetentionPeriods["user-1"] = 0
	cfgProvider.UserRetentionPeriods["user-2"] = 0

	require.NoError(t, cleaner.cleanUsers(ctx, false))
	assertBlocksExist(true, true, true, true)

	assert.NoError(t, prom_testutil.GatherAndCompare(reg, strings.NewReader(`
		# HELP cortex_bucket_blocks_count Total number of blocks in the bucket. Includes blocks marked for deletion, but not partial blocks.
		# TYPE cortex_bucket_blocks_count gauge
		cortex_bucket_blocks_count{user="user-1"} 2
		cortex_bucket_blocks_count{user="user-2"} 2
		# HELP cortex_bucket_blocks_marked_for_deletion_count Total number of blocks marked for deletion in the bucket.
		# TYPE cortex_bucket_blocks_marked_for_deletion_count gauge
		cortex_bucket_blocks_marked_for_deletion_count{user="user-1"} 0
		cortex_bucket_blocks_marked_for_deletion_count{user="user-2"} 0
		# HELP cortex_compactor_blocks_marked_for_deletion_total Total number of blocks marked for deletion in compactor.
		# TYPE cortex_compactor_blocks_marked_for_deletion_total counter
		cortex_compactor_blocks_marked_for_deletion_total{reason="retention"} 0
	`),
		"cortex_bucket_blocks_count",
		"cortex_bucket_blocks_marked_for_deletion_count",
		"cortex_compactor_blocks_marked_for_deletion_total",
	))
}
```